docs/security/index.md

```diff
@@ -0,0 +1,453 @@
+to focus on their implementation detail.
+
+### `Entitlements<?>`
+
+`Entitlements` is a generic type that holds user entitlements, it holds the `encoded()` byte sequence as a `byte[]`, a
+`schema()` identifier as a `short`, and its `decodedEntitlements()` method provides access to the concrete entitlements
+type a security plugin uses for its implementation.  Again the `AbstractSecurityPrimitive` class provides a basic
+implementation of most of the interface.
+
+### `SecurityPlugin`
+
+`SecurityPlugin` is the top level entrypoint API by which applications interact with the Security Plugin API.  Primarily
+this is just a place for applications to obtain instances of the other interfaces they need to carry out data processing
+and access control tasks.
+
+An implementation should generally make the instances of the other interfaces singletons where the API allows it e.g.
+there should be a single shared `SecurityLabelsParser`.  However implementations **MUST** be aware that they will be
+used in multi-threaded environments thus any singleton instance they return **MUST** be thread-safe.
+
+Note that some of the interfaces a plugin provides instances of have specific lifecycles associated with them, e.g.
+`Authorizer`, so for those an implementation **MUST** ensure they respect the defined lifecycle of that interface.
+These interfaces are clearly marked by having them extend `AutoCloseable` so that the Java compiler will issue warnings
+if application developers are not actively `close()`'ing the obtained instances, or using them in a `try-with-resources`
+block to implicitly `close()` them as shown in the [Example Usage](#example-usage).
+
+### `IdentityProvider`
+
+The Core Platform relies upon Open ID Connect (OIDC) as the authentication mechanism and uses Bearer tokens in the form
+of cryptographically signed JSON Web Tokens (JWTs).  Depending on how a given deployment is configured there *MAY* be
+multiple underlying identity providers, but by design individual applications don't need to be aware of these, they
+merely need to be able to cryptographically verify the presented tokens using the public keys of the identity providers.
+
+However, in practical terms most applications deal with users, rather than tokens, when making access decisions.  Thus a
+key part of a plugin is the `IdentityProvider`.  This interface translates from the authenticated users JWT into a user
+identifier that the application can use, typically this is some form of username/email address.  However, depending on
+the security plugin and/or deployment this may need to be a more opaque identifier since email addresses may be reused
+across multiple identity providers e.g. some combination of the `iss` and `sub` claims from the JWT.
+
+For most plugins we don't expect the mapping of JWT to user identity to vary so we provide a generic
+`DefaultIdentityProvider` that we expect most plugins will reuse.  This maps the user identity by looking for claims in
```

Comment:
Suggest the claim used as the principal's identifier should be fixed in configuration rather than fall back through a list of preferences.

Reply:
Clarified that this is fully configurable, and added a note below stating that a customer could fix this to a specific claim if they so desired.